src/java/org/apache/cassandra/io/sstable/format/bti/PartitionIndex.java

```diff
@@ -0,0 +1,452 @@
+                // force a read of all the pages of the index
+                for (long pos = 0; pos < fh.dataLength(); pos += PageAware.PAGE_SIZE)
+                {
+                    rdr.seek(pos);
+                    csum += rdr.readByte();
+                }
+                logger.trace("Checksum {}", csum);      // Note: trace is required so that reads aren't optimized away.
+            }
+
+            return new PartitionIndex(fh, root, keyCount, first, last);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        fh.close();
+    }
+
+    @Override
+    public Throwable close(Throwable accumulate)
+    {
+        return fh.close(accumulate);
+    }
+
+    public Reader openReader()
+    {
+        return new Reader(this);
+    }
+
+    protected IndexPosIterator allKeysIterator()
+    {
+        return new IndexPosIterator(this);
+    }
+
+    protected Rebufferer instantiateRebufferer()
+    {
+        return fh.instantiateRebufferer(null);
+    }
+
+
+    /**
+     * @return the file handle to the file on disk. This is needed for locking the index in RAM.
+     */
+    FileHandle getFileHandle()
+    {
+        return fh;
+    }
+
+    private static long getIndexPos(ByteBuffer contents, int payloadPos, int bytes)
+    {
+        if (bytes > 7)
+        {
+            ++payloadPos;
+            bytes -= 7;
+        }
+        if (bytes == 0)
+            return NOT_FOUND;
+        return SizedInts.read(contents, payloadPos, bytes);
+    }
+
+    public interface Acceptor<ArgType, ResultType>
+    {
+        ResultType accept(long position, boolean assumeNoMatch, ArgType v) throws IOException;
+    }
+
+    /**
+     * Provides methods to read the partition index trie.
+     * Thread-unsafe, uses class members to store lookup state.
+     */
+    public static class Reader extends Walker<Reader>
+    {
+        protected Reader(PartitionIndex index)
+        {
+            super(index.instantiateRebufferer(), index.root);
+        }
+
+        /**
+         * Finds a candidate for an exact key search. Returns an ifile (if positive) or dfile (if negative, using ~)
+         * position. The position returned has a low chance of being a different entry, but only if the sought key
+         * is not present in the file.
+         */
+        public long exactCandidate(DecoratedKey key)
+        {
+            // A hit must be a prefix of the byte-comparable representation of the key.
+            int b = follow(key);
+            // If the prefix ended in a node with children it is only acceptable if it is a full match.
+            if (b != ByteSource.END_OF_STREAM && hasChildren())
+                return NOT_FOUND;
+            if (!checkHashBits(key.filterHashLowerBits()))
+                return NOT_FOUND;
+            return getCurrentIndexPos();
+        }
+
+        final boolean checkHashBits(short hashBits)
+        {
+            int bytes = payloadFlags();
+            if (bytes <= 7)
```

Comment:
Is it possible for bytes to be anything other than `0` or `> 7`, given we use `nonZeroSize()`? Is it just equivalent to this?

```
return bytes > 7 && buf.get(payloadPosition()) == (byte) hashBits;
```